dot/parachain/collator-protocol/validator_side.go

```diff
@@ -115,12 +122,323 @@ func (cpvs CollatorProtocolValidatorSide) handleNetworkEvents(event network.Netw
+	// download blocks etc to reach the tip of the chain faster.
+	var majorSyncing bool
+
+	liveHeads := cpvs.liveHeads[:]
+	liveHeads = append(liveHeads, parachaintypes.ActivatedLeaf{
+		Hash:   signal.Activated.Hash,
+		Number: signal.Activated.Number,
+	})
+
+	newLiveHeads := []parachaintypes.ActivatedLeaf{}
+
+	for _, head := range liveHeads {
+		var sliceContains bool
+		for _, deactivated := range signal.Deactivated {
+			if head.Hash == deactivated {
+				sliceContains = true
+				break
+			}
+		}
+		if sliceContains {
+			newLiveHeads = append(newLiveHeads, head)
+		}
+	}
+
+	sort.Sort(SortableActivatedLeaves(newLiveHeads))
+	// TODO: do I need to store these live heads or just pass them to update view?
+	cpvs.liveHeads = newLiveHeads
+
+	if !majorSyncing {
+		// update our view
+		err := cpvs.updateOurView(liveHeads)
+		if err != nil {
+			logger.Errorf("updating our view: %w", err)
+		}
+	}
 }
 
-func (cpvs CollatorProtocolValidatorSide) ProcessBlockFinalizedSignal() {
-	// NOTE: nothing to do here
+func (cpvs *CollatorProtocolValidatorSide) updateOurView(liveHeads []parachaintypes.ActivatedLeaf) error {
+	headHashes := []common.Hash{}
+	for _, head := range cpvs.liveHeads {
+		headHashes = append(headHashes, head.Hash)
+	}
+	newView := View{
+		heads:           headHashes,
+		finalizedNumber: cpvs.finalizedNumber,
+	}
+
+	if cpvs.localView == nil {
+		*cpvs.localView = newView
+		return nil
+	}
+
+	if cpvs.localView.checkHeadsEqual(newView) {
+		// nothing to update
+		return nil
+	}
+
+	*cpvs.localView = newView
+
+	// TODO: send ViewUpdate to all the collation peers and validation peers (v1, v2, v3)
+	// https://github.com/paritytech/polkadot-sdk/blob/aa68ea58f389c2aa4eefab4bf7bc7b787dd56580/polkadot/node/network/bridge/src/rx/mod.rs#L969-L1013
+
+	// TODO: Create our view and send collation events to all subsystems about our view change
+	// Just create the network bridge and do both of these tasks as part of those. That's the only way it makes sense.
+
+	err := cpvs.handleOurViewChange(newView)
+	if err != nil {
+		return fmt.Errorf("handling our view change: %w", err)
+	}
+	return nil
+}
+
+func (cpvs *CollatorProtocolValidatorSide) handleOurViewChange(view View) error {
+
+	// a lot is happenning here.
+
+	// 1. Find out removed leaves (hashes) and newly added leaves
+	// 2. Go over each new leaves,
+	// - check if perspective parachain mode is enabled
+	// - assign incoming
+	// - insert active leaves and per relay parent
+	activeLeaves := cpvs.activeLeaves
+
+	removed := []common.Hash{}
+	for activeLeaf := range activeLeaves {
+		if !slices.Contains(view.heads, activeLeaf) {
+			removed = append(removed, activeLeaf)
+		}
+	}
+
+	newlyAdded := []common.Hash{}
+	for _, head := range view.heads {
+		if _, ok := activeLeaves[head]; !ok {
+			newlyAdded = append(newlyAdded, head)
+		}
+	}
+
+	// handled newly added leaves
+	for _, leaf := range newlyAdded {
+		mode := perspectiveParchainMode(leaf)
```

Comment:
I think it is `prospectiveParachainMode`, right?
```suggestion
		mode := prospectiveParachainMode(leaf)
```